File: scripts/templategenerator/TemplateLoader.py
from .Template import Template
from .TemplateType import TemplateType
from os import listdir, path
import re


class TemplateLoader:
    _TEMPLATE_REGEX = re.compile(
        '^(?:[a-zA-Z\\\\:-]*schemas\\\\)'
        '(?:project|player|shared)\\\\'
        '(?P<type>fragment|query|mutation)\\\\'
        '(?:[a-zA-Z]+)\\.gql$'
    )

    def __init__(self, resource_dir):
        self.operations = dict()
        self._fragments = dict()
        self._resource_dir_path = resource_dir

    def load(self) -> None:
        self._load_raw_resources()

    def _load_raw_resources(self) -> None:
        if self._resource_dir_path is None or path.exists(self._resource_dir_path):
            pass

        path_stack = list()
        path_stack.append(self._resource_dir_path)

        while len(path_stack) > 0:
            resource_path = path_stack.pop()

            if path.isdir(resource_path):
                files = listdir(resource_path)
                if len(files) > 0:
                    path_stack += ['%s\\%s' % (resource_path, f) for f in files]
            else:
                matcher = re.match(TemplateLoader._TEMPLATE_REGEX, path.abspath(resource_path))
                if matcher is None:
                    continue

                group_type = matcher.groupdict()['type']
                for template_type in TemplateType:
                    if template_type.name.lower() == group_type.lower():
                        self._load_and_cache_template_contents(resource_path, template_type)
                        break

    def _load_and_cache_template_contents(self, file_path: str, template_type: TemplateType) -> None:
        with open(file_path, 'r') as f:
            contents = f.readlines()

        if contents is None:
            return None

        namespace = Template.read_namespace(contents)
        if namespace is None:
            return None

        if template_type == TemplateType.FRAGMENT:
            self._fragments[namespace] = Template(namespace, template_type, contents, self._fragments)
        elif template_type == TemplateType.MUTATION or template_type == TemplateType.QUERY:
            self.operations[namespace] = Template(namespace, template_type, contents, self._fragments)
        else:
            raise ValueError('Unknown template type: %s' % template_type)

File: scripts/templategenerator/TemplateLoader.cpp
/* Copyright 2021 Enjin Pte. Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "TemplateLoader.hpp"

#include "enjinsdk_utils/StringUtils.hpp"
#include <fstream>
#include <stack>
#include <stdexcept>
#include <vector>

namespace fs = std::filesystem;

namespace enjin::generator {

TemplateLoader::TemplateLoader(fs::path resource_dir)
        : resource_dir(std::move(resource_dir)),
          fragments(std::make_shared<std::map<std::string, Template>>()) {
}

void TemplateLoader::load() {
    load_raw_resources();
}

void TemplateLoader::load_raw_resources() {
    if (!fs::exists(resource_dir)) {
        return;
    }

    // Selects the appropriate regular expression based on the filesystem
    std::regex re = fs::path::preferred_separator == '/'
            ? std::regex(FORWARD_SLASH_REGEX)
            : std::regex(BACK_SLASH_REGEX);

    std::stack<fs::path> path_stack;
    path_stack.push(resource_dir);

    while (!path_stack.empty()) {
        auto resource_path = path_stack.top();
        path_stack.pop();

        if (fs::is_directory(resource_path)) {
            for (const auto& entry : fs::directory_iterator(resource_path)) {
                path_stack.push(entry.path());
            }
        } else {
            std::string abs_path(fs::absolute(resource_path).string());
            std::smatch matcher;

            // Check if any matches were made
            if (!std::regex_search(abs_path, matcher, re)) {
                continue;
            }

            auto group_type = matcher[1];
            for (const auto& type : TemplateType::values()) {
                if (utils::to_lower(type.to_string()) == utils::to_lower(group_type)) {
                    load_and_cache_template_contents(resource_path, type);
                    break;
                }
            }
        }
    }
}

void TemplateLoader::load_and_cache_template_contents(const std::filesystem::path& path, TemplateType type) {
    std::ifstream file(path);

    std::vector<std::string> contents;
    std::string line;
    while (std::getline(file, line)) {
        contents.push_back(line);
    }

    file.close();

    if (contents.empty()) {
        return;
    }

    auto t_namespace = Template::read_namespace(contents);
    if (!t_namespace.has_value()) {
        return;
    }

    switch (type.value()) {
        case TemplateType::Value::Fragment:
            fragments->emplace(t_namespace.value(), Template(t_namespace.value(), type, contents, fragments));
            break;
        case TemplateType::Value::Mutation:
        case TemplateType::Value::Query:
            operations.emplace(t_namespace.value(), Template(t_namespace.value(), type, contents, fragments));
            break;
        default:
            throw std::runtime_error("Unknown template type: " + type.to_string());
    }
}

const std::map<std::string, Template>& TemplateLoader::get_operations() const {
    return operations;
}

}
